### ISSUE TITLE
gumtree implementation in swift programming language

### ISSUE THREAD

**Issue by aminul7506** [GumTreeDiff/gumtree#164]
How can I implement gumtree for swift programming language using the existing codebase ? Should I implement a swift AST in java for this ?

**Comment by jrfaller** [comment]
Hi! It would be a nice feature! There are two solution, depending on if a parser for swift exists in Java. In this case you can use the parsers API to generate an AST compatible with Gumtree (see how it is done for instance in the module gen.js that uses Rhino. If this is not the case, you'll have to  build an external tool that produces a XML file compatible with Gumtree, this how it's done in the gen.c module for instance. Which parser do you want to use?

**Comment by aminul7506** [comment]
A parser for swift does not exist in Java but we have a parser for swift written in swift from https://github.com/yanagiba/swift-ast . I want to use this AST for swift. For this what should I do ? Should I make a tool from this swift AST that produces gumtree compatible XML file ? It would be helpful for me if you tell me about this.

**Comment by jrfaller** [comment]
Yes you have to make an executable tool that uses your parser to produce a XML compatible with GumTree!

You can take a look to the format here (in testSerializeTree()): https://github.com/GumTreeDiff/gumtree/blob/develop/core/src/test/java/com/github/gumtreediff/test/TestTreeIoUtils.java

You can look how to implement the adapter here : https://github.com/GumTreeDiff/gumtree/tree/develop/gen.c/src/main/java/com/github/gumtreediff/gen/c

Good luck with that!

**Comment by aminul7506** [comment]
Thank you sir. I am trying this.

**Comment by aminul7506** [comment]
Sir, can you please tell me what is Type, Label , Length and Position of a Node with an example?

**Comment by jrfaller** [comment]
For instance consider the Python code `v = 12;`

We have the following tree : 

<tree type="Module" pos="0" length="7" >
	<tree type="Assign" pos="0" length="7">
		<tree type="NameStore" label="v" pos="0" length="1" />
		<tree type="Num" label="12" pos="4" length="2" />
	</tree>
</tree>

Therefore, **type**: lexical category of the node
**value** (only valid for leaf nodes): associated value (for instance v for the NameStore or 12 for the Num
**pos**: absolute position of the begining of the node in the text file
**length**: length of the node in the text file, in characters (we have end_pos = pos + length)

Cheers!

**Comment by aminul7506** [comment]
Another question. Where I have to install cgum for seeing the diff for C programme?

**Comment by aminul7506** [comment]
Another question. Where I have to install cgum for seeing the diff for C programme? I am trying to see the c programe diff in gumtree but failed

**Comment by jrfaller** [comment]
Hi! cgum is here : https://github.com/GumTreeDiff/cgum

But if it's easier you can also look into the gen.python and https://github.com/GumTreeDiff/pythonparser tools that does exactly the same also. I just fixed these two yesterday so they use all the good practices for implementing an external generator.

Cheers!

**Comment by jrfaller** [comment]
Hi! Any progress on this issue? Cheers!

**Comment by jrfaller** [comment]
Hi @aminul7506!

If you are still interested by having Swift into GumTree I think I can use the tree sitter parser easily now?
